Funções e módulos
=====================

Introdução
------------

As funções nos permitem agrupar várias declarações em um bloco lógico. Nos comunicamos com uma função através de uma interface claramente definida, fornecendo certos parâmetros para a função e recebendo algumas informações de volta. Além dessa interface, geralmente não sabemos como exatamente uma função faz o trabalho para obter o valor que retorna

Por exemplo, a função `math.sqrt`: não sabemos como exatamente ela calcula a raiz quadrada, mas sabemos sobre a interface: se passarmos `x` para a função, ela retornará (uma aproximação de) $ \\sqrt{x} $.

Esta abstração é uma coisa útil: é uma técnica comum na engenharia para quebrar um sistema em componentes menores (caixa preta) que trabalham juntos através de interfaces bem definidas, mas que não precisam saber sobre as realizações internas da funcionalidade um do outro. Na verdade, não ter que se preocupar com esses detalhes de implementação pode ajudar a ter uma visão mais clara do sistema composto por muitas dessas componentes.

As funções fornecem os blocos básicos de funcionalidade em programas maiores (e simulações computacionais) e ajudam a controlar a complexidade inerente ao processo.

Podemos agrupar funções em um módulo Python (veja [módulos](#Modulos)) e, assim, criar nossas próprias bibliotecas de funcionalidades.

Usando funções
---------------

A palavra "função" tem diferentes significados em matemática e programação. Na programação, ela se refere a uma sequência de operações que executam uma computação. Por exemplo, a função `sqrt()`, definida no módulo de matemática `math`, calcula a raiz quadrada de um determinado valor:

In [1]:
from math import sqrt
sqrt(4)

2.0

O valor que passamos para a função `sqrt` é 4 neste exemplo. Esse valor é chamado de *argumento* da função. Uma função pode ter mais de um argumento.

A função retorna o valor 2.0 (o resultado de sua computação) para o "contexto de chamada". Esse valor é chamado de *valor de retorno* da função.

É comum dizer que uma função *leva* um argumento e *retorna* um resultado ou valor de retorno.

#### Confusão comum sobre a impressão e os valores de retorno 

É um erro comum do principiante confundir a *impressão* de valores com *valores de retorno*. No exemplo a seguir, é difícil ver se a função `math.sin` retorna um valor ou se ela imprime o valor:

In [2]:
import math
math.sin(2)

0.9092974268256817

Nós importamos o módulo `math` e chamamos a função `math.sin` com o argumento `2`. A chamada `math.sin(2)` realmente *retornará* o valor `0.909...`, não o imprimirá. No entanto, como não foi atribuído o valor de retorno a uma variável, o prompt do Python imprimirá o objeto retornado.

A seguinte sequência alternativa funciona apenas se o valor for retornado:

In [3]:
x = math.sin(2)
print(x)

0.9092974268256817


O valor de retorno da chamada da função `math.sin (2)` é atribuído à variável `x` e `x` é impresso na próxima linha.

Geralmente, as funções devem executar "silenciosamente" (ou seja, não imprimir nada) e reportar o resultado de sua computação através do valor de retorno.

Parte da confusão sobre valores impressos versus valores de retorno no prompt do Python vem da impressão rápida do Python (uma representação) dos objetos retornados *se* os objetos retornados não são atribuídos. Geralmente, ver os objetos retornados é exatamente o que queremos (como normalmente nos preocupamos com o objeto retornado), apenas ao aprender Python. Isso pode causar confusão leve sobre funções que retornam valores ou valores de impressão.

##### Informação adicional

- *Think Python* apresenta uma introdução gentil às funções (em que se baseia o parágrafo anterior) no [capítulo 3 (Funções)] (http://www.greenteapress.com/thinkpython/html/book004.html) e [capítulo 6 ( Funções frutíferas)] (http://www.greenteapress.com/thinkpython/html/book007.html). (Vide: http://novatec.com.br/livros/pense-em-python/")

Definindo funções
------------------

Formato genérico de definição de uma função:

```python
def minha_funcao(arg1, arg2, ..., argn):
    """Docstring opcional"""

    # Implementacao da funcao

    return resultado  # opcional

# isto nao eh parte da funcao
algum_comando
```

For example, the function `greeting` will print “Hello World” when called (and is fruitless as it does not return a value).

A terminologia de Allen Downey (em seu livro [Think Python](http://www.greenteapress.com/thinkpython/html/index.html)) de funções frutíferas e infrutíferas distingue funções que retornam um valor daquelas que não retornam valor algum. A distinção refere-se a se uma função fornece um valor de retorno (=frutífera) ou se a função não retorna explicitamente um valor (=infrutífera). Se as funções não usam a instrução `return`, tendemos a dizer que a função não retorna nada (enquanto que, na realidade, sempre retornará o objeto `None` quando ele termina - mesmo que nela falte a instrução `return`).

Por exemplo, a função `cumprimento` imprimirá "Ola, mundo!" quando chamada (e é infrutífero, pois não retorna um valor).

In [1]:
def cumprimento():
    print("Ola, mundo!")

Se chamarmos essa função:

In [2]:
cumprimento()

Ola, mundo!


ela imprime “Ola, Mundo!” para `stdout`, com esperado. Se atribuírmos o valor de retorno da função para a variável `x`, podemos inspecioná-lo na sequencia:

In [3]:
x = cumprimento()

Ola, mundo!


In [4]:
print(x)

None


e encontramos que a função `cumprimento`, de fato, retornou, o objeto `None`.

Outro exemplo para uma função que não retorna valor algum (isso significa que não há palavra-chave `return` na função) seria:

In [5]:
def printpluses(n): 
    print(n * "+")

Geralmente, funções que retornam valores são mais úteis, pois estas podem ser usadas para montar código (talvez como outra função) combinando-as de maneira "esperta"". Vejamos alguns exemplos de funções que retornam um valor.

Suponhamos que precisemos definir uma função que calcule o quadrado de uma determinada variável. O código-fonte da função poderia ser:

In [6]:
def quadrado(x):
    return x * x

A palavra-chave `def` diz ao Python que estamos *definindo* uma função naquele ponto. A função leva um argumento (`x`). A função retorna `x*x`, que é, claramento, $ x^2 $. Aqui está um exemplo que mostra como a função pode ser definida e usada:

In [7]:
def quadrado(x):
    return x * x

for i in range(5):
    i_quadrado = quadrado(i)
    print(i, '*', i, '=', i_quadrado)

0 * 0 = 0
1 * 1 = 1
2 * 2 = 4
3 * 3 = 9
4 * 4 = 16


Vale mencionar que as linhas 1 e 2 definem a função `quadrado` ao passo que as linhas 4 a 6 são o programa principal.

Podemos definir funções que levam mais do que um argumento:

In [8]:
import math

def hipotenusa(x, y):
    return math.sqrt(x * x + y * y)

Também é possível retornar mais de um argumento. Aqui está um exemplo de uma função que converte uma determinada *string* com todos os caracteres em letras maiúsculas e todos os caracteres em letras minúsculas e retorna as duas versões. Incluímos o programa principal para mostrar como esta função pode ser chamada:

In [9]:
def maiusculasEMinusculas(string):
    return string.upper(), string.lower()

palavra = 'Banana'

uppercase, lowercase = maiusculasEMinusculas(palavra)

print(palavra, 'em minusculas:', lowercase,
      'e em maiusculas', uppercase)

Banana em minusculas: banana e em maiusculas BANANA


Podemos definir múltiplas funções em Python em um arquivo. Aqui está um exemplo com duas funções:

In [11]:
def retorna_estrelas( n ):
    return n * '*'

def imprime_centrado_estrelas( string ):
    linelength = 46 
    starstring = retorna_estrelas((linelength - len(string)) // 2)

    print(starstring + string + starstring)

imprime_centrado_estrelas('Ola, Mundo!')

*****************Ola, Mundo!*****************


##### Leitura complementar

-   [Python Tutorial: Seção 4.6 Definindo Funções](http://docs.python.org/tutorial/controlflow.html#defining-functions)

Default values and optional parameters
--------------------------------------

Python allows to define *default* values for function parameters. Here is an example: This program will print the following output when executed: So how does it work? The function `print_mult_table` takes two arguments: `n` and `upto`. The first argument `n` is a “normal” variable. The second argument `upto` has a default value of 10. In other words: should the user of this function only provide one argument, then this provides the value for `n` and `upto` will default to 10. If two arguments are provided, the first one will be for `n` and the second for `upto` (as shown in the code example above).

Modules
-------

Modules

-   Group together functionality

-   Provide namespaces

-   Python’s standard library contains a vast collection of modules - “Batteries Included”

  -   Try `help(’modules’)`

-   Means of extending Python

### Importing modules

In [14]:
import math

This will introduce the name `math` into the namespace in which the import command was issued. The names within the `math` module will not appear in the enclosing namespace: they must be accessed through the name `math`. For example: `math.sin`.

In [15]:
import math, cmath

More than one module can be imported in the same statement, although the [Python Style Guide](http://www.python.org/dev/peps/pep-0008/) recommends not to do this. Instead, we should write

In [16]:
import math
import cmath

import math as mathematics

The name by which the module is known locally can be different from its “official” name. Typical uses of this are

-   To avoid name clashes with existing names

-   To change the name to something more manageable. For example `import SimpleHTTPServer as shs`. This is discouraged for production code (as longer meaningful names make programs far more understandable than short cryptic ones), but for interactively testing out ideas, being able to use a short synonym can make your life much easier. Given that (imported) modules are first class objects, you can, of course, simply do `shs = SimpleHTTPServer` in order to obtain the more easily typable handle on the module.

<!-- -->

In [17]:
from math import sin

This will import the `sin` function from the `math` module, but it will not introduce the name math into the current namespace. It will only introduce the name `sin` into the current namespace. It is possible to pull in more than one name from the module in one go:

In [18]:
from math import sin, cos

Finally, let’s look at this notation:

In [19]:
from math import *

Once again, this does not introduce the name math into the current namespace. It does however introduce *all public names* of the math module into the current namespace. Broadly speaking, it is a bad idea to do this:

-   Lots of new names will be dumped into the current namespace.

-   Are you sure they will not clobber any names already present?

-   It will be very difficult to trace where these names came from

-   Having said that, some modules (including ones in the standard library, recommend that they be imported in this way). Use with caution!

-   This is fine for interactive quick and dirty testing or small calculations.

### Creating modules

A module is in principle nothing else than a python file. Here is an example of a module file which is saved in `module1.py`:

```python
def someusefulfunction():
    pass

print("My name is", __name__)
```

We can execute this (module) file as a normal python program (for example `python module1.py`):

In [20]:
cd code/

/Users/fangohr/hg/teaching-python/book/text/code


In [21]:
!python module1.py

My name is __main__


We note that the Python magic variable `__name__` takes the value `__main__` if the program file `module1.py` is executed.

On the other hand, we can *import* `module1.py` in another file (which could have the name `prog.py`), for example like this:

In [22]:
import module1            #in file prog.py

My name is module1


When Python comes across the `import module1` statement in `prog.py`, it looks for the file `module1.py` in the current working directory (and if it can’t find it there in all the directories in `sys.path`) and opens the file `module1.py`. While parsing the file `module1.py` from top to bottom, it will add any function definitions in this file into the `module1` name space in the calling context (that is the main program in `prog.py`). It this example, there is only the function `someusefulfunction`. Once the import process is completed, we can make use of `module1.someusefulfunction` in `prog.py`. If Python comes across statements other than function (and class) definitions while importing `module1.py`, it carries those out immediately. In this case, it will thus come across the statement `print(My name is, __name__)`.

Note the difference to the output if we *import* `module1.py` rather than executing it on its own: `__name__` inside a module takes the value of the module name if the file is imported.

### Use of \_\_name\_\_

In summary,

-   `__name__` is `__main__` if the module file is run on its own

-   `__name__` is the name of the module (i.e. the module filename without the `.py` suffix) if the module file is imported.

We can therefor use the following `if` statement in `module1.py` to write code that is *only run* when the module is executed on its own: This is useful to keep test programs or demonstrations of the abilities of a module in this “conditional” main program. It is common practice for any module files to have such a conditional main program which demonstrates its capabilities.

### Example 1

The next example shows a main program for the another file `vectools.py` that is used to demonstrate the capabilities of the functions defined in that file:

```python
from __future__ import division
import math

import numpy as N


def norm(x):
    """returns the magnitude of a vector x"""
    return math.sqrt(sum(x ** 2))


def unitvector(x):
    """returns a unit vector x/|x|. x needs to be a numpy array."""
    xnorm = norm(x)
    if xnorm == 0:
        raise ValueError("Can't normalise vector with length 0")
    return x / norm(x)


if __name__ == "__main__":
    #a little demo of how the functions in this module can be used:
    x1 = N.array([0, 1, 2])
    print("The norm of " + str(x1) + " is " + str(norm(x1)) + ".")
    print("The unitvector in direction of " + str(x1) + " is " \
        + str(unitvector(x1)) + ".")
```

If this file is executed using `python vectools.py`, then `__name__==__main__` is true, and the output reads

In [23]:
!python3 vectortoolswithconditionalmain.py

The norm of [0 1 2] is 2.23606797749979.
The unitvector in direction of [0 1 2] is [ 0.          0.4472136   0.89442719].


If this file is imported (i.e. used as a module) into another python file, then `__name__==__main__` is false, and that statement block will not be executed (and no output produced).

This is quite a common way to conditionally execute code in files providing library-like functions. The code that is executed if the file is run on its own, often consists of a series of tests (to check that the file’s functions carry out the right operations – *regression tests* or *unit tests* ), or some examples of how the library functions in the file can be used.

### Example 2

Even if a Python program is not intended to be used as a module file, it is good practice to always use a conditional main program:

-   often, it turns out later that functions in the file can be reused (and saves work then)

-   this is convenient for regression testing.

Suppose an exercise is given to write a function that returns the first 5 prime numbers, and in addition to print them. (There is of course a trivial solution to this as we know the prime numbers, and we should imagine that the required calculation is more complex). One might be tempted to write

In [24]:
def primes5():
    return (2, 3, 5, 7, 11)

for p in primes5():
    print("%d" % p, end=' ')

2 3 5 7 11 

It is better style to use a conditional main function, i.e.:

In [25]:
def primes5():
    return (2, 3, 5, 7, 11)

if __name__=="__main__":
    for p in primes5():
        print("%d" % p, end=' ')

2 3 5 7 11 

A purist might argue that the following is even cleaner:

In [26]:
def primes5():
    return (2, 3, 5, 7, 11)

def main():
    for p in primes5():
        print("%d" % p, end=' ')

if __name__=="__main__":
    main()

2 3 5 7 11 

but either of the last two options is good.

The example in [Many ways to compute a series](#Many-ways-to-compute-a-series) demonstrates this technique. Including functions with names starting with `test_` is compatible with the very useful py.test regression testing framework (see <http://pytest.org/>).

#### Further Reading

-   [Python Tutorial Section 6](http://docs.python.org/tutorial/modules.html#modules)